src/frontend/components/Chat/SharedChatComponents/InputDisplayNameOverlay/index.spec.tsx

```diff
@@ -92,7 +98,150 @@ describe('<InputDisplayNameOverlay />', () => {
     expect(mockSetOverlay).not.toHaveBeenCalled();
   });
 
+  it('enters a valid nickname but the server takes too long to answer.', async () => {
+    const mockConverse =
+      converse.claimNewNicknameInChatRoom as jest.MockedFunction<
+        typeof converse.claimNewNicknameInChatRoom
+      >;
+
+    mockConverse.mockImplementation(
+      async (
+        _displayName: string,
+        _callbackSuccess: () => void,
+        callbackError: (stanza: Nullable<HTMLElement>) => void,
+      ) => {
+        await new Promise((r) => setTimeout(r, 2000));
+        callbackError(null);
+      },
+    );
+
+    render(
+      wrapInIntlProvider(
+        <InputDisplayNameOverlay setOverlay={mockSetOverlay} />,
+      ),
+    );
+    const inputTextbox = screen.getByRole('textbox');
+    const validateButton = screen.getByRole('button');
+    userEvent.type(inputTextbox, 'John_Doe');
+    expect(validateButton.querySelector('svg')).toBeTruthy();
+    act(() => userEvent.click(validateButton));
+    expect(mockSetOverlay).toHaveBeenCalledTimes(0);
+    expect(converse.claimNewNicknameInChatRoom).toHaveBeenCalledTimes(1);
+    expect(validateButton.querySelector('svg')).toBeNull();
+    await act(async () => {
+      jest.advanceTimersByTime(3000);
+    });
+    expect(validateButton.querySelector('svg')).toBeTruthy();
+    screen.getByText('The server took too long to respond. Please retry.');
+    expect(useChatItemState.getState().displayName).toEqual(null);
+  });
+
+  it('enters a valid nickname but it is already used in the chat', async () => {
+    const mockConverse =
+      converse.claimNewNicknameInChatRoom as jest.MockedFunction<
+        typeof converse.claimNewNicknameInChatRoom
+      >;
+
+    mockConverse.mockImplementation(
+      async (
+        _displayName: string,
+        _callbackSuccess: () => void,
+        callbackError: (stanza: Nullable<HTMLElement>) => void,
+      ) => {
+        await new Promise((r) => setTimeout(r, 2000));
+        const parser = new DOMParser();
+        callbackError(
+          parser.parseFromString(
+            '<error code="409" />',
+            'text/xml',
+          ) as any as HTMLElement,
+        );
+      },
+    );
+
+    render(
+      wrapInIntlProvider(
+        <InputDisplayNameOverlay setOverlay={mockSetOverlay} />,
+      ),
+    );
+    const inputTextbox = screen.getByRole('textbox');
+    const validateButton = screen.getByRole('button');
+    userEvent.type(inputTextbox, 'John_Doe');
+    expect(validateButton.querySelector('svg')).toBeTruthy();
+    act(() => userEvent.click(validateButton));
+    expect(mockSetOverlay).toHaveBeenCalledTimes(0);
+    expect(converse.claimNewNicknameInChatRoom).toHaveBeenCalledTimes(1);
+    expect(validateButton.querySelector('svg')).toBeNull();
+    await act(async () => {
+      jest.advanceTimersByTime(3000);
+    });
```

Comment:
Is it needed ? You are not testing the timeout but the username usage.

Reply:
Yes, because when the component is waiting the prosody's answer a spinner is displayed in place of the button' svg, and this test is checking this mechanism too.